lib/erb2epp/rewriter.rb

```diff
@@ -0,0 +1,215 @@
+      end
+      res
+    end
+
+    # Store local variable names to prepend with '$' later
+    # Look for "var =" or "(var1, var2) ="
+    def collect_local_vars(tokens)
+      allowed_tokens = " -\n(,)=".chars.freeze
+      maybe_vars = []
+      tokens.each do |type, value|
+        case type
+        when :on_ident
+          maybe_vars << value
+        else
+          break unless allowed_tokens.include? value # Not an evaluation
+
+          if value == '='
+            @local_vars.concat maybe_vars
+            break
+          end
+        end
+      end
+    end
+
+    # Rewrite local variables
+    def rewrite_local_vars(tokens)
+      res = []
+      tokens.each do |type, value|
+        case type
+        when :on_ident
+          value = "$#{value}" if @local_vars.include? value
+        end
+        res << [type, value]
+      end
+      res
+    end
+
+    # Add an opening curly bracked to the end of `if` statement line
+    def rewrite_if(tokens)
+      return tokens unless tokens.count { |x| x[1] == '{' }.zero?
+
+      # We need the opening bracket
+      ocb_pos = tokens.size - 1
+      # Go back until non-[ -\n] found
+      allowed_tokens = " -\n".chars.freeze
+      ocb_pos -= 1 while allowed_tokens.include? tokens[ocb_pos][1]
+
+      res = tokens[..ocb_pos]
+      res << [:on_sp, ' ']
+      res << [:on_kw, '{']
+      res.concat tokens[ocb_pos + 1..]
+      res
+    end
+
+    # Swap opening curly bracket and block vars: {|x| ..} => |x| {..}
+    def rewrite_blockvars(tokens)
+      pos = { lbrace: 0, lpipe: 0, rpipe: 0 }
+      res = []
+      idx = 0
+      tokens.each do |token|
+        _, value = token
+        case value
+        when '{'
+          pos[:lbrace] = idx
+        when '|'
+          pos[(pos[:lpipe].positive? ? :rpipe : :lpipe)] = idx if pos[:lbrace].positive?
+        end
+        res << token
+        idx += 1
+        next unless pos.values.all?(&:positive?)
+
+        # When every position is positive then we found the whole "{ |...|" string, time to rewrite it
+        new_res = res[0..(pos[:lbrace]) - 1]                             # copy everything before lbrace
+        new_res << [:on_sp, ' '] if new_res.last[0] != :on_sp            # add space if none
+        new_res.concat collect_vars_in_pipes(res[(pos[:lpipe])..(pos[:rpipe])]) # from lpipe to rpipe
+        new_res << [:on_sp, ' ']     # space before lbrace
+        new_res << [:on_lbrace, '{'] # lbrace
+
+        pos = { lbrace: 0, lpipe: 0, rpipe: 0 }
+        res = new_res
+        idx = res.length
+      end
+      res
+    end
+
+    # Rewrite a piece of Ruby code
+    def rewrite_code(code)
+      tokens = rewrite_tokens(code)
+      rewritten_code = tokens.map { |x| x[1] }.join
+
+      tokens = rewrite_if(tokens) if /^[- ]*if/.match? rewritten_code
+      tokens = rewrite_blockvars(tokens) if /{[^}]*?\|[^\|]*?\|/.match? rewritten_code
+
+      collect_local_vars(tokens) if /[a-z][A-Za-z0-9_(), ]*=/.match? rewritten_code
```

Comment:
Why do you look at the rewritten code instead of looking at the tokens?

Reply:
You're right.. `rewritten_code` is a workaround to use regexps instead of crawling tokens.. I cut some edges here, but ideally tokens should be used instead.

Reply:
Just wrapped this with a `tokens_match?()` function which converts tokes to a string to match it against the regexp specified.. if you have a better idea please let me know 🙇🏻